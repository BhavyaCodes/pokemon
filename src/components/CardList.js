import React, { useState } from "react";
import pokemonList from "../pokemon.json";
import PokeCard from "./PokeCard";
import Pagination from "./pagination";

function CardList({ searchQuery }) {
  const [currentPage, setCurrentPage] = useState(1);
  const [pageSize] = useState(9);

  /**
   * This function sorts pokemons by their ids in ascending order
   */

  const pokemonComparator = (firstPokemon, secondPokemon) => {
    /**
     * Math.sign() return -1, 0 or 1 *only* for any signed integer
     */
    return Math.sign(firstPokemon.id - secondPokemon.id);
  };

  /**
   * Following a functional programming principle here which states that
   * we should not mutate objects in any way.
   *
   * Refer this : https://www.learningjournal.guru/article/scala/functional-programming/immutability-in-functional-programming/
   */
  const sortedPokemonList = pokemonList.sort(pokemonComparator);
  const filteredSortedPokemonList = sortedPokemonList.filter((pokemon) =>
    pokemon.name.toLowerCase().includes(searchQuery.toLowerCase())
  );

  return (
    <React.Fragment>
      <div className="row row-cols-1 row-cols-md-3 mt-3">
        {filteredSortedPokemonList
          .slice((currentPage - 1) * pageSize, currentPage * pageSize)
          .map((pokemon) => (
            <PokeCard key={pokemon.id} pokemon={pokemon} />
          ))}
      </div>
      <Pagination
        currentPage={currentPage}
        setCurrentPage={setCurrentPage}
        pageSize={pageSize}
        totalData={filteredSortedPokemonList.length}
      />
    </React.Fragment>
  );
}

export default CardList;
