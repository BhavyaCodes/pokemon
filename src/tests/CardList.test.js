import React from "react";
import { render } from "@testing-library/react";

import CardList from "../components/CardList";

it("renders CardList component without crashing", () => {
  const { asFragment } = render(<CardList searchQuery={"pika"} />);
  expect(asFragment()).toMatchSnapshot();
});
