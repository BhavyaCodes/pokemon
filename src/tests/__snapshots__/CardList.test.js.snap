// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders CardList component without crashing 1`] = `
<DocumentFragment>
  <div
    class="row row-cols-1 row-cols-md-3 mt-3"
  >
    <div
      class="col mb-4"
    >
      <div
        class="card"
        tabindex="0"
      >
        <img
          alt="pikachu"
          class="card-img-top"
          src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/25.png"
        />
        <div
          class="card-body"
        >
          <h5
            class="card-title text-center"
          >
            pikachu
          </h5>
          <p
            class="card-text text-center"
          >
            <span
              class="badge badge-pill badge-electric mx-1"
              data-testid="pokemon-type-electric"
            >
              electric
            </span>
          </p>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
